k4Reco/DDPlanarDigi/components/DDPlanarDigi.cpp

```diff
@@ -0,0 +1,348 @@
+
+  m_histograms[hitE].reset(new Gaudi::Accumulators::Histogram<1>{this, "hitE", "hitEnergy in keV", {1000, 0, 200}});
+  m_histograms[hitsAccepted].reset(
+      new Gaudi::Accumulators::Histogram<1>{this, "hitsAccepted", "Fraction of accepted hits [%]", {201, 0, 100.5}});
+
+  m_geoSvc = serviceLocator()->service(m_geoSvcName);
+}
+
+StatusCode DDPlanarDigi::initialize() {
+  const auto detector = m_geoSvc->getDetector();
+
+  const auto         surfMan = detector->extension<dd4hep::rec::SurfaceManager>();
+  dd4hep::DetElement det     = detector->detector(m_subDetName.value());
+  surfaceMap                 = surfMan->map(m_subDetName.value());
+
+  if (!surfaceMap) {
+    std::stringstream err;
+    err << " Could not find surface map for detector: " << det.name() << " in SurfaceManager ";
+    throw std::runtime_error(err.str());
+  }
+
+  // Get and store the name for a debug message later
+  (void)this->getProperty("SimTrackerHitCollectionName", m_collName);
+
+  return StatusCode::SUCCESS;
+}
+
+std::tuple<TrackerHitPlaneColl, Association> DDPlanarDigi::operator()(
+    const SimTrackerHitCollection& simTrackerHits, const Header& headers) const {
+  auto seed = m_uidSvc->getUniqueID(headers[0].getEventNumber(), headers[0].getRunNumber(), this->name());
+  info() << "Using seed " << seed << " for event " << headers[0].getEventNumber() << " and run "
+         << headers[0].getRunNumber() << endmsg;
+  m_engine.seed(seed);
+  // For rng calls, use the fact that drawing from a
+  // Gaussian with mean mu and variance sigma^2 is the same as drawing from
+  // a normal distribution and then multiplying by sigma and adding mu
+  auto dist = std::normal_distribution<double>(0, 1);
+
+  int nCreatedHits   = 0;
+  int nDismissedHits = 0;
+
+  auto trkhitVec = edm4hep::TrackerHitPlaneCollection();
+  auto thsthcol  = edm4hep::MCRecoTrackerHitPlaneAssociationCollection();
+
+  std::string cellIDEncodingString = m_geoSvc->constantAsString(m_encodingStringVariable.value());
+  dd4hep::DDSegmentation::BitFieldCoder bitFieldCoder(cellIDEncodingString);
+
+  int nSimHits = simTrackerHits.size();
+  debug() << "Processing collection " << m_collName << " with " << simTrackerHits.size() << " hits ... " << endmsg;
+
+  for (const auto& hit : simTrackerHits) {
+    ++(*m_histograms[hitE])[hit.getEDep() * (dd4hep::GeV / dd4hep::keV)];
+
+    if (hit.getEDep() < m_minEnergy) {
+      debug() << "Hit with insufficient energy " << hit.getEDep() * (dd4hep::GeV / dd4hep::keV) << " keV" << endmsg;
+      continue;
+    }
+
+    const int cellID0 = hit.getCellID();
+
+    // get the measurement surface for this hit using the CellID
+    dd4hep::rec::SurfaceMap::const_iterator sI = surfaceMap->find(cellID0);
+
+    if (sI == surfaceMap->end()) {
+      std::stringstream err;
+      err << " DDPlanarDigi::processEvent(): no surface found for cellID : " << std::endl;
+      // <<   cellid_decoder( hit ).valueString() ;
+      throw std::runtime_error(err.str());
```

Comment:
Gaudi brings fmtlib, right? Since c++20 will (once all compilers have implemented it), will also have `std::format` we could start using that functionality instead of creating strings this way.

Reply:
Using `std::format` since we're building with C++20

Reply:
Do the compiler versions we use already have support for that? I think gcc has that only from 13 onwards and clang has partial support from 14 and full from 17 onwards:

https://en.cppreference.com/w/cpp/compiler_support/20

Reply:
Ah indeed, it's one of the latest ones. I changed it to use `fmtlib`